app/src/renderer/vuex/modules/transactions.js

```diff
@@ -0,0 +1,123 @@
+let { uniqBy } = require("lodash")
+
+export default ({ node }) => {
+  let state = {
+    loading: false,
+    wallet: [], // {height, result: { gas, tags }, tx: { type, value: { fee: { amount: [{denom, amount}], gas}, msg: {type, inputs, outputs}}, signatures} }}
+    staking: []
+  }
+
+  let mutations = {
+    setWalletTxs(state, txs) {
+      state.wallet = txs
+    },
+    setStakingTxs(state, txs) {
+      state.staking = txs
+    },
+    setHistoryLoading(state, loading) {
+      state.loading = loading
+    },
+    setTransactionTime(state, { scope, blockHeight, blockMetaInfo }) {
+      state[scope] = state[scope].map(t => {
+        if (t.height === blockHeight) {
+          t.time = blockMetaInfo && blockMetaInfo.header.time
+        }
+        return t
+      })
+    }
+  }
+
+  let actions = {
+    async getAllTxs({ commit, dispatch }) {
+      commit("setHistoryLoading", true)
+      const stakingTxs = await dispatch("getTx", "staking")
+      commit("setStakingTxs", stakingTxs)
+
+      const walletTxs = await dispatch("getTx", "wallet")
+      commit("setWalletTxs", walletTxs)
+
+      const allTxs = stakingTxs.concat(walletTxs)
+      await dispatch("enrichTransactions", {
+        transactions: allTxs,
+        scopes: ["staking", "wallet"]
+      })
+      commit("setHistoryLoading", false)
+    },
+    // after creating getAllTxs, getStakingTxs and getWalletTxs are no longer
+    // referenced anywhere. keeping them here in case there is need for them later
+    // async getStakingTxs({ commit, dispatch, rootState }) {
+    //   commit("setHistoryLoading", true)
+    //   const transactions = await dispatch("getTx", "staking")
+    //   commit("setStakingTxs", transactions)
+    //
+    //   await dispatch("enrichTransactions", {
+    //     transactions,
+    //     scopes: ["staking"]
+    //   })
+    //   commit("setHistoryLoading", false)
+    // },
+    // async getWalletTxs({ rootState, commit, dispatch }) {
+    //   commit("setHistoryLoading", true)
+    //   const transactions = await dispatch("getTx", "wallet")
+    //   commit("setStakingTxs", transactions)
+    //
+    //   await dispatch("enrichTransactions", {
+    //     transactions,
+    //     scopes: ["wallet"]
+    //   })
+    //   commit("setHistoryLoading", false)
+    // },
+    async getTx({ rootState }, type) {
+      let address = rootState.user.address
+      let res
+      switch (type) {
+        case "staking":
+          res = await node.getDelegatorTxs(address)
+          break
+        case "wallet":
+          res = await node.txs(rootState.user.address)
+          break
+        default:
+          throw new Error("Unknown transaction type")
+      }
+      if (!res) return []
+      res = res.map(r => {
+        r.type = type
+        return r
+      })
+      return uniqBy(res, "hash")
+    },
```

Comment:
See https://www.nodeguy.com/p/87512fce-e9ec-4402-8f18-7e914d93a3d7/

```JavaScript
const fp = require(`lodash/fp`)

async getTx(
  {
    rootState: {
      user: { address }
    }
  },
  type
) {
  let response

  switch (type) {
    case "staking":
      response = await node.getDelegatorTxs(address)
      break
    case "wallet":
      response = await node.txs(address)
      break
    default:
      throw new Error("Unknown transaction type")
  }

  return response ? uniqBy(response.map(fp.set(`type`, type)), "hash") : []
}
```

Reply:
short and functional, but I had my problems reading this. how about making it a bit more descriptive:

```
const enrichedTransactions = response.map(fp.set(`type`, type))
return response ? uniqBy(enrichedTransactions, "hash") : []
```

Apart from that: Maybe let's avoid demanding refactors of copy pasted code as it bloats PRs.

Reply:
woa i didn't realize you could set default arguments as object children!

Reply:
@faboweb I like your improvement.

> Apart from that: Maybe let's avoid demanding refactors of copy pasted code as it bloats PRs.

I don't want to demand refactors (I'm still figuring out how to use GitHub properly here) but I don't see how the source of the code makes any difference—it should be improved regardless.

> woa i didn't realize you could set default arguments as object children!

This is [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring).

Reply:
@NodeGuy valid point. I add an agenda point for tomorrows meeting to discuss a strategy.